pkg/syncer/cnsoperator/controller/cnsnodevmattachment/cnsnodevmattachment_controller.go

```diff
@@ -140,9 +143,36 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 
 	backOffDuration = make(map[string]time.Duration)
 
+	// Predicates are used to determine under which conditions
+	// the reconcile callback will be made for an instance.
+	pred := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return true
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			oldObj, ok := e.ObjectOld.(*cnsnodevmattachmentv1alpha1.CnsNodeVmAttachment)
+			if oldObj == nil || !ok {
+				return false
+			}
+			newObj, ok := e.ObjectNew.(*cnsnodevmattachmentv1alpha1.CnsNodeVmAttachment)
+			if newObj == nil || !ok {
+				return false
+			}
+			// Return true if spec or status has changed.
```

Comment:
Not sure if returning true on status change is going to fix the issue I observed. In the case I observed, thread-1 updated the status to "Attached" but there was an immediate reconciliation that had a stale CnsNodeVmAttachment that had a status of "NotAttached" or nil.

Reply:
I believe it will fix the issue. The case you observed would have the flow:

1. thread-1 adds the finalizer to the CR. <This kicks in another reconciliation in thread-2>

2. thread-2 reads the CR With finalizer added but status `NotAttached`.

3. thread-1 attaches the volume and updates the status to `Attached`. <This also kicks in reconciliation but we return seeing the status has already been Attached>

4. thread-2 tries to attach the volume. Observes ResourceInUse fault but checks the disk is attached to the same VM. So it tries to update the status in its instance copy from `NotAttached` to `Attached`. But the etcd already has the instance status as `Attached`. So this update is rejected with `IsConflict` error and we report an attach failure.

----------

With the fix provided, the flow would be as follows:

1. thread-1 adds the finalizer to the CR. <This will not kick in reconciliation in another thread>

2. thread-1 attaches the volume and updates the status to `Attached`. <This also kicks in reconciliation but we return seeing the status has already been Attached>

Reply:
> With the fix provided, the flow would be as follows:
> 
> thread-1 adds the finalizer to the CR. [This will not kick in reconciliation in another thread]
> 
> thread-1 attaches the volume and updates the status to Attached. [This also kicks in reconciliation but we return seeing the status has already been Attached]

The issue I observed is that after thread-1 attaches the volume and updates the status to `Attached`, another reconciliation kicks in. This new thread reads a stale CnsNodeVmAttachment(**_this is the problem_**) that has status set to nil and tries to attach the same PV to the VM. And since it had a stale CnsNodeVmAttachment, the update on the CR also fails with `the object has been modified; please apply your changes to the latest version and try again`. Thats the reason why I said that this code change does not fix the issue I observed. Hope this detail clarifies.